Wrappers/Python/cil/optimisation/functions/IndicatorBox.py

```diff
@@ -29,67 +35,389 @@ class IndicatorBox(Function):
+    
+    def convex_conjugate(self,x):
+        '''Convex conjugate of IndicatorBox at x'''
+        return x.maximum(0).sum()   
             
-            self.proximal(x/tau, tau, out=out)
-            out *= -1*tau
-            out += x
+    def _proximal(self, outarr):
+        np.clip(outarr, None if self.orig_lower is None else self.lower, 
+                        None if self.orig_upper is None else self.upper, out=outarr)
+
+
+## Utilities
+def _get_as_nparray_or_number(x):
+    '''Returns x as a numpy array or a number'''
+    try:
+        return x.as_array()
+    except AttributeError:
+        # In this case we trust that it will be either a numpy ndarray 
+        # or a number as described in the docstring
+        return x
+
+@numba.jit(nopython=True, parallel=True)
+def _array_within_limits_ff(x, lower, upper, breaking):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    arr = x.ravel()
+    for i in numba.prange(x.size):
+        j = numba.np.ufunc.parallel._get_thread_id()
+        
+        if breaking[j] == 0 and (arr[i] < lower or arr[i] > upper):
+            breaking[j] = 1
+
+@numba.jit(nopython=True, parallel=True)
+def _array_within_limits_af(x, lower, upper, breaking):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != lower.size:
+        raise ValueError('x and lower must have the same size')
+    arr = x.ravel()
+    loarr = lower.ravel()
+    for i in numba.prange(x.size):
+        j = numba.np.ufunc.parallel._get_thread_id()
+        
+        if breaking[j] == 0 and (arr[i] < loarr[i] or arr[i] > upper):
+            breaking[j] = 1
+
+@numba.jit(parallel=True, nopython=True)
+def _array_within_limits_aa(x, lower, upper, breaking):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != lower.size or x.size != upper.size:
+        raise ValueError('x, lower and upper must have the same size')
+    arr = x.ravel()
+    uparr = upper.ravel()
+    loarr = lower.ravel()
+    for i in numba.prange(x.size):
+        j = numba.np.ufunc.parallel._get_thread_id()
+        
+        if breaking[j] == 0 and (arr[i] < loarr[i] or arr[i] > uparr[i]):
+            breaking[j] = 1
+
+@numba.jit(nopython=True, parallel=True)
+def _array_within_limits_fa(x, lower, upper, breaking):
+    '''Returns 0 if all elements of x are within [lower, upper]'''
+    if x.size != upper.size:
+        raise ValueError('x and upper must have the same size')
+    arr = x.ravel()
+    uparr = upper.ravel()
+    for i in numba.prange(x.size):
+        j = numba.np.ufunc.parallel._get_thread_id()
+        
+        if breaking[j] == 0 and (arr[i] < lower or arr[i] > uparr[i]):
+            breaking[j] = 1
+
+##########################################################################
+
+@numba.jit(nopython=True, parallel=True)
+def _proximal_aa(x, lower, upper):
+    '''Similar to np.clip except that the clipping range can be defined by ndarrays'''
```

Comment:
the clipping range can be defined as ndarrays with np.clip. Isn't it just a performance gain?

Reply:
yes you are right